x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -216,8 +218,42 @@ private void searchGenericEvents(
         ActionListener<GetStackTracesResponse> submitListener,
         GetStackTracesResponseBuilder responseBuilder
     ) {
-        responseBuilder.setSamplingRate(1.0d);
-        client.prepareSearch(request.indices())
+        StopWatch watch = new StopWatch("getSamplingRate");
+        client.prepareSearch(request.getIndices())
+            .setSize(0)
+            .setTrackTotalHits(true)
+            .setRequestCache(true)
+            .setPreference(String.valueOf(request.hashCode()))
+            .setQuery(request.getQuery())
+            .execute(ActionListener.wrap(searchResponse -> {
+                long sampleCount = searchResponse.getHits().getTotalHits().value;
+                int requestedSampleCount = request.getSampleSize();
+                // random sampler aggregation does not support sampling rates between 0.5 and 1.0 -> clamp to 1.0
+                if (sampleCount <= requestedSampleCount * 2L) {
```

Comment:
Having these values hard-coded isn't ideal. Can we somehow read these values out from the liable component? Or change that component that it automatically uses 1.0 when the samplingRate is in the unsupported area?

Reply:
Yeah, I'm not a fan of hard-coded values either. However, in this case I believe it's appropriate. The fact that random sampler disallows probabilities between (0.5;1) is just [documented knowledge](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-random-sampler-aggregation.html) but is not retrievable from the code (i.e. hardcoded there as well):

https://github.com/elastic/elasticsearch/blob/72fe8b30d5bc69b69e3af0c0a9e6478f362a6f34/server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregationBuilder.java#L58-L60

Basically this code is here to adhere to the preconditions of the called code and I'd argue that having the constant embedded here is ok.